gax/src/main/java/com/google/api/gax/rpc/StubSettings.java

```diff
@@ -334,6 +371,19 @@ public B setClock(ApiClock clock) {
 
     public B setEndpoint(String endpoint) {
       this.endpoint = endpoint;
+      if (this.endpoint != null && this.mtlsEndpoint == null) {
+        this.mtlsEndpoint = this.endpoint.replace("googleapis.com", "mtls.googleapis.com");
+      }
+      return self();
+    }
+
+    protected B setOverridableEndpoint(String endpoint) {
```

Comment:
If this method is explicitly called from the generated client, it will make newly generated clients incompatible with older versions of gax, which may cause breakages. Please explore the possibilities of making new generated clients compatible with older versions of gax (mtls feature is not expected to work in such case, but everything else should still work).

Reply:
How widespread are the gax-java dependencies? Understanding the scenarios in which newly-generated clients could use an older gax-java version would help us explore mitigation approaches.

Reply:
Generated clients have a pom.xml file which defines the gax-java version to use, when the user installs the client the right version of gax-java will be installed. I don't think this will be a problem. I think backward compatibility is more about a new version of gax-java doesn't break an old version client lib.

It doesn't look possible to make newly-generated client compatible with the old gax-java. The reason is that in the following function in the generated client stub settings, I need to somehow tell gax-java that the endpoint is set by the client, so gax-java can override it with mtls endpoint if some client cert conditions meet. Since both the user and the client use `setEndpoint`, there is no way I can tell the difference without changing the existing surfaces or introducing new surfaces. Therefore in gax-java I introduced a `setOverridableEndpoint` function which can only be called by the clients, and in the following client code, I just need to change `setEndpoint` to `setOverridableEndpoint`.

```
private static Builder createDefault() {
      Builder builder = new Builder((ClientContext) null);
      builder.setTransportChannelProvider(defaultTransportChannelProvider());
      builder.setCredentialsProvider(defaultCredentialsProviderBuilder().build());
      builder.setInternalHeaderProvider(defaultApiClientHeaderProviderBuilder().build());
      builder.setEndpoint(getDefaultEndpoint());
      return initDefaults(builder);
}
```